From be30bf4427454a62346d8a1f03c276fec4dc1fa6 Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Tue, 7 Apr 2026 04:03:09 +0800
Subject: [PATCH 95/95] BACKPORT: ifcvt: reject use of ctrap post reload [PR
 105192]

The ctrap expanders of several targets can end up creating pseudos.  And
it would be really counter-intuitive to disallow a define_expand to
generate pseudos.  So don't invoke the expander post reload.

	PR target/105192

gcc/

	* ifcvt.cc (find_if_header): Don't attempt to use ctrap expander
	after reload.

gcc/testsuite/

	* gcc.c-torture/compile/pr105192.c: New test.

(cherry picked from commit 7f1e96ee025ab481cf0893b3a4ce82909b22d8f9)
(cherry picked from commit a26a9a2665ba10b2ecc2413c69c29cf96e0d5908)
---
 gcc/ifcvt.cc                                   | 2 +-
 gcc/testsuite/gcc.c-torture/compile/pr105192.c | 9 +++++++++
 2 files changed, 10 insertions(+), 1 deletion(-)
 create mode 100644 gcc/testsuite/gcc.c-torture/compile/pr105192.c

diff --git a/gcc/ifcvt.cc b/gcc/ifcvt.cc
index e4168d95ad9..372ab4db7f4 100644
--- a/gcc/ifcvt.cc
+++ b/gcc/ifcvt.cc
@@ -4738,7 +4738,7 @@ find_if_header (basic_block test_bb, int pass)
       && cond_exec_find_if_block (&ce_info))
     goto success;
 
-  if (targetm.have_trap ()
+  if (!reload_completed && targetm.have_trap ()
       && optab_handler (ctrap_optab, word_mode) != CODE_FOR_nothing
       && find_cond_trap (test_bb, then_edge, else_edge))
     goto success;
diff --git a/gcc/testsuite/gcc.c-torture/compile/pr105192.c b/gcc/testsuite/gcc.c-torture/compile/pr105192.c
new file mode 100644
index 00000000000..0cc61505218
--- /dev/null
+++ b/gcc/testsuite/gcc.c-torture/compile/pr105192.c
@@ -0,0 +1,9 @@
+/* { dg-additional-options "-fno-if-conversion -fharden-compares" } */
+
+int a, b;
+
+void
+foo (void)
+{
+  b = a % 0 > 0;
+}
-- 
2.54.0

